### ISSUE TITLE
Search algorithm related to steps

### ISSUE THREAD

**Issue by viperfx** [AHAAAAAAA/PokemonGo-Map#442]
Hi, 

Could someone explain the search algorithm a bit more? Does it make a search that is a box of the steps from the centre? or is the search space more of a radius? If I wanted to search frequently in a small space, how should I cover the other cells nearby efficiently with multiple instances? 

Thanks,

**Comment by jangrewe** [comment]
i think it's not steps as in "foot steps", but in "iteration steps to cover an area"

1 step = an area with 1 cell, ~200-250m wide and high
3 steps = an area with 9 cells, which is 3 cells wide and 3 cells high (A1, A2, A3, B1, B2, B3, C1, C2, C3)
5 steps = 25 cells, 5 wide / 5 high (A1...A5, ..., E1...E5)

**Comment by viperfx** [comment]
Oh yep, I am aware they related to a distance. The wiki says it is 5 steps is roughly 1.2km. So I guess it is a box that it covers then?

**Comment by Skullmonkey** [comment]
It was kinda explained in #211 even though that's not what the OP wanted. I know there was a 'wandering' bug that I think still persists in the most recent version (can't test because I cant seem to get past the 'no-len payload' error right now) that keeps moving the center point in one direction after a complete scan.

TL;DR: Square from center

**Comment by jangrewe** [comment]
hmm, if 5 steps is a radius of 1.2km (let's take 1.25km), the the diameter is 2.5km, which would make every cell be 500m wide/high (not 200-250m, as assumed above).

so with 3 steps, you'd have one cell in the center and then one cell to each side and in each corner.
if we're going with 500m as the cell size, you'd set the next location to 1.5km from the first location (2x 1 1/2 cells - or 750m (250m+500m) to the border of the first instance, and another 750m from the border of the next instance to its center location)

at least that's what it works out to in my head ;-)

**Comment by AHAAAAAAA** [comment]
[Here's the algorithm.](https://gist.github.com/yigitserin/0fdaea1a6c6b5412f87781950ab164e1)

[It looks like this, each location by which step it is.](https://i.imgur.com/I1IVitb.png)

**Comment by Plop** [comment]
Can you explain why I can see the gyms but not the pokestop or pokemon within steps ? If I set 8 steps, I know there is an gym and a pokestop near it, but I only see the gym, even though the pokestop is "in" the square.
I don't really get how it works. And I will have more data at the south of the center than at the west. For example, I will see pokemon and gyms at the south, at great distance, but closer to the center, at west, nothing. I'm not sure if I'm clear in what I say...

**Comment by NgJason** [comment]
> set_location_coords(x * 0.0025 + origin_lat, y * 0.0025 + origin_lon, 0)

Similar to @jangrewe , I think this line tells us exactly how the search algorithm works. Given an offset of cells, by dx,dy -> x , y , it adds 0.0025*x + lat or 0.0025*y + long. 
Assuming that is correct, I believe those are the new CENTER coordinates that are sent over to the Pokemon servers. 

Even though this mimic a SQUARE "cell" iteration, I don't think the visible Pokemons are returned in a SQUARE fashion. Ninatic most likely are doing this by: Given a set of coordinate (Center of a circle // var center), return all Pokemons within this Area(center). This is why sometimes you will see the same Pokemon twice if you print from the console. There are overlapping areas in this algorithm (Theoretical assumption):

![theory](https://cloud.githubusercontent.com/assets/6984333/16969193/ff585a6e-4dc7-11e6-9e21-5929771695a9.png)

**Comment by jangrewe** [comment]
Thanks, that's exactly the picture i had in mind (except the circular radius), but my Paint skills are not on fleek :grin:

**Comment by viperfx** [comment]
Awesome, thanks guys!

**Comment by macgngsta** [comment]
are you sure the search algorithm is accurate?

on line 11, you square maxSize to be 100, but never store it back into maxSize
so on line 12, it reads as `-5 < x <= 5 and -5 < y <= 5`
shouldn't it be `-50 < x <= 50 and -50 < y <= 50`

btw we visited this issue earlier #249

**Comment by magnusvmt** [comment]
no I'm pretty sure you want it to go 10 steps in both the x and y direction to create a square (not a perfect square since 1km longitude is not the same degreechange as 1km in latitude) . The algorithm seems good to me.

**Comment by Plop** [comment]
At the south, you can see there are the stops and gyms, but no pokemon, Same problem at the extremity of the square, in fact. I tried yesterday to go there and check if there were pokemons, and there were; But they don't show up on the map :/ And this doesn't really look like a square to me, more like a rectangle.
Maybe I'm wrong ?
![image](https://cloud.githubusercontent.com/assets/876470/16979481/a4a14f6e-4e61-11e6-97a0-c8713a04d1f3.png)

**Comment by magnusvmt** [comment]
The reason it's not a square is probably because 0.0025 long != 0.0025 lat unless you're at the equator.

**Comment by Plop** [comment]
Hmm, alright.

But I tried with [this](https://github.com/AHAAAAAAA/PokemonGo-Map/issues/494#issuecomment-233909853) in order to see the precise markers, and I'm wondering why I can see the gym and pokestops at the south (Noyal sur Vilaine), but not at the west (at the right of "La Jarsais). it's kind of bothering :/ Is there a technical reason for this ?


![](http://i.imgur.com/xGvrSNg.png)

**Comment by magnusvmt** [comment]
I'm not sure, but if you look at NgJasons image it's possible that the gym is right in between 4 "circles".

**Comment by NgJason** [comment]
Hi @Plop , @magnusvmt is refering to my post at: https://github.com/AHAAAAAAA/PokemonGo-Map/issues/442#issuecomment-233789595

Magnusvmt might be right that a gym could appear in between 4 circles. 
![ff585a6e-4dc7-11e6-9e21-5929771695a9](https://cloud.githubusercontent.com/assets/6984333/17000695/85e78b62-4e78-11e6-9cd0-e1f528cae5ae.png)

You could try the following:
Take smaller "steps"
distance = 0.002 #keep playing with this, lower the number, closer the circles will be. Keep in mind that this will increase the overlapping area due the the properties of circles.

set_location_coords(x * distance + origin_lat, y * distance  + origin_lon, 0)

![ff585a6e-4dc7-11e6-9e21-5929771695a9](https://cloud.githubusercontent.com/assets/6984333/17000932/be761a42-4e79-11e6-8cec-c53181b43625.png)

Ignoring the overlapping and assuming Nintanic uses circles (2d spheres), we can try to find the "best" distance that leaves no "gray area" untreated (TREATING OVERLAPPING AREAS as neither good nor bad) by finding the largest width ( W ) of a square possible in the given circle ( determined by Nintanic back end server). 

Let W be Width of a square , and R as the radius set by Nintanic (unknown), we can calculate the ideal W by using the Pythagorean Theroem:

W^2 + W^2 = R^2. 
W = sqr root[ (R^2)/2] or
R = sqr root [2*(W^2)].

**Comment by Plop** [comment]
I'm not sure about the distance setting in the example.py file, but I will try to do something with this.
Thanks !

**Comment by NgJason** [comment]
@Plop just a note, the original line is this:
set_location_coords(x * 0.0025 + origin_lat, y * 0.0025 + origin_lon, 0)

new code could be:
distance = 0.002
set_location_coords(x * distance + origin_lat, y * distance + origin_lon, 0)

**Comment by Plop** [comment]
Oh, sorry, I wasn't looking for the correct line. Thanks, I will give it a try :)

**Comment by Plop** [comment]
Well, I've been trying to lower the distance and increase the number of steps, but I still had the same problem with the same gym. The gym at the east is detected right away, from far away, but the gym at the center is never correctly detected, unless the marker is really on it.
![image](https://cloud.githubusercontent.com/assets/876470/17015209/2d53b3ac-4f29-11e6-92a6-5b8053b5506a.png)

After trying to set the location at the left of the gym, it is correctly detected right away. 

![image](https://cloud.githubusercontent.com/assets/876470/17015268/7f40ea86-4f29-11e6-8117-e016d0b3c473.png)

But if I try several locations at the right : 
![image](https://cloud.githubusercontent.com/assets/876470/17015409/5d20a404-4f2a-11e6-9156-89817d5e1c58.png)



I tried with another map : 
![image](https://cloud.githubusercontent.com/assets/876470/17015342/fbdad160-4f29-11e6-89f2-c9878384d95e.png)

I really don't understand what's going on there. I tried with smaller steps, many steps, and it seems that it is able to detect gyms from far away, so why not this one ?
